lib/std/tar.zig

```diff
@@ -1,172 +1,1096 @@
-            // This code does not look at the mode bits yet. To implement this feature,
-            // the implementation must be adjusted to look at the mode, and check the
-            // user executable bit, then call fchmod on newly created files when
-            // the executable bit is supposed to be set.
-            // It also needs to properly deal with ACLs on Windows.
-            @panic("TODO: unimplemented: tar ModeMode.executable_bit_only");
-        },
+const V7HeaderDummy = [257]u8;
+
+const UstarHeader = extern struct {
+    v7_header: V7HeaderDummy,
+    magic: [6]u8, // 257..263
+    version: [2]u8, // 263..265
+    user_name: [32]u8, // 265..297
+    group_name: [32]u8, // 297..329
+    dev_major: [8]u8, // 329..337
+    dev_minor: [8]u8, // 337..345
+    filename_prefix: [155]u8, // 345..500
+    __padding: [12]u8,
+
+    pub fn magicVersion(ustar: UstarHeader) u64 {
+        return mem.readIntBig(u64, @ptrCast([*]const u8, &ustar.magic)[0..8]);
+    }
+
+    comptime {
+        assert(@sizeOf(UstarHeader) == block_len);
+    }
+};
+
+const StarHeader = extern struct {
+    v7_header: V7HeaderDummy,
+    magic: [6]u8, // 257..263
+    version: [2]u8, // 263..265
+    user_name: [32]u8, // 265..297
+    group_name: [32]u8, // 297..329
+    dev_major: [8]u8, // 329..337
+    dev_minor: [8]u8, // 337..345
+    filename_prefix: [131]u8, // 345..476
+    access_time: [12]u8, // 476..488
+    change_time: [12]u8, // 488..500
+    __padding: [8]u8,
+    trailer: [4]u8, // 508..512
+
+    comptime {
+        assert(@sizeOf(StarHeader) == block_len);
+    }
+};
+
+const GnuHeader = extern struct {
+    v7_header: V7HeaderDummy,
+    magic: [6]u8, // 257..263
+    version: [2]u8, // 263..265
+    user_name: [32]u8, // 265..297
+    group_name: [32]u8, // 297..329
+    dev_major: [8]u8, // 329..337
+    dev_minor: [8]u8, // 337..345
+    access_time: [12]u8, // 345..357
+    change_time: [12]u8, // 357..369
+    __padding: [17]u8,
+    sparse: [24 * 4 + 1]u8, // 386..483
+    real_size: [12]u8, // 483..495
+    __padding2: [17]u8,
+
+    comptime {
+        assert(@sizeOf(GnuHeader) == block_len);
     }
-    var file_name_buffer: [255]u8 = undefined;
-    var buffer: [512 * 8]u8 = undefined;
-    var start: usize = 0;
-    var end: usize = 0;
-    header: while (true) {
-        if (buffer.len - start < 1024) {
-            std.mem.copy(u8, &buffer, buffer[start..end]);
-            end -= start;
-            start = 0;
+};
+
+pub fn headerIterator(
+    reader: anytype,
+    buf: *[512 * 2]u8,
+    arena: mem.Allocator,
+) HeaderIterator(@TypeOf(reader)) {
+    return HeaderIterator(@TypeOf(reader)){
+        .reader = reader,
+        .buf = buf[0..block_len],
+        .pax_buf = buf[block_len .. 2 * block_len],
+        .arena = arena,
+    };
+}
+
+pub fn HeaderIterator(comptime Reader: type) type {
+    return struct {
+        reader: Reader,
+        buf: Block = undefined,
+        pax_buf: Block = undefined,
```

Comment:
This has the same issue as the longname and longlink, it needs to be allocated because it can be >512 bytes.

Reply:
Woops! Thank you. :+1:

Reply:
addressed just now in forced push

Reply:
glad you highlighted this. `Block` is actually defined as `*[512]u8`. The `= undefined` default was a remnant and has been removed now.

Reply:
addressed. i just rebased on latest master. i'm going to mark conversations previous to this point as resolved.